python/semantic_kernel/utils/tracing/decorators.py

```diff
@@ -0,0 +1,233 @@
+#
+# Code to trace model activities with the OTel semantic conventions.
+# This code contains experimental features and may change in the future.
+# To enable these features, set one of the following senvironment variables to true:
+#    SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS
+#    SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS_SENSITIVE
+
+import functools
+import json
+import os
+from collections.abc import Callable
+from typing import Any
+
+from opentelemetry import trace
+from opentelemetry.trace import Span, StatusCode
+
+from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+from semantic_kernel.contents.chat_history import ChatHistory
+from semantic_kernel.contents.chat_message_content import ITEM_TYPES, ChatMessageContent
+from semantic_kernel.contents.function_call_content import FunctionCallContent
+from semantic_kernel.utils.tracing.const import (
+    COMPLETION_EVENT,
+    COMPLETION_EVENT_COMPLETION,
+    COMPLETION_TOKEN,
+    ERROR_TYPE,
+    FINISH_REASON,
+    MAX_TOKEN,
+    MODEL,
+    OPERATION,
+    PROMPT_EVENT,
+    PROMPT_EVENT_PROMPT,
+    PROMPT_TOKEN,
+    RESPONSE_ID,
+    SYSTEM,
+    TEMPERATURE,
+    TOP_P,
+)
+
+OTEL_ENABLED_ENV_VAR = "SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS"
+OTEL_SENSITIVE_ENABLED_ENV_VAR = "SEMANTICKERNEL_EXPERIMENTAL_GENAI_ENABLE_OTEL_DIAGNOSTICS_SENSITIVE"
+
+
+_enable_diagnostics = os.getenv(OTEL_ENABLED_ENV_VAR, "false").lower() in ("true", "1", "t")
+_enable_sensitive_events = os.getenv(OTEL_SENSITIVE_ENABLED_ENV_VAR, "false").lower() in ("true", "1", "t")
+
+# Creates a tracer from the global tracer provider
+tracer = trace.get_tracer(__name__)
+
+
+def are_model_diagnostics_enabled() -> bool:
+    """Check if model diagnostics are enabled.
+
+    Model diagnostics are enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set.
+    """
+    return _enable_diagnostics or _enable_sensitive_events
+
+
+def are_sensitive_events_enabled() -> bool:
+    """Check if sensitive events are enabled.
+
+    Sensitive events are enabled if EnableSensitiveEvents is set.
+    """
+    return _enable_sensitive_events
+
+
+def trace_chat_completion(model_provider: str) -> Callable:
+    """Decorator to trace chat completion activities."""
+
+    def inner_trace_chat_completion(completion_func: Callable) -> Callable:
+        @functools.wraps(completion_func)
+        async def wrapper_decorator(*args: Any, **kwargs: Any) -> list[ChatMessageContent]:
+            chat_history: ChatHistory = kwargs["chat_history"]
+            settings: PromptExecutionSettings = kwargs["settings"]
+
+            if hasattr(settings, "ai_model_id") and settings.ai_model_id:
+                model_name = settings.ai_model_id
+            elif hasattr(args[0], "ai_model_id") and args[0].ai_model_id:
+                model_name = args[0].ai_model_id
+            else:
+                model_name = "unknown"
```

Comment:
I am not sure if this is the right way to get the model id. Not all prompt execution settings will have the `ai_model_id` attribute. I think this brings up a question about the `ai_model_id` attr in some of the prompt execution settings and the one in `AIServiceClientBase`. Why do need two?

Reply:
It's tricky because for some, like Azure OpenAI the ai_model_id is the deployment name, so that is a single thing, while for OpenAI you could theoretically override that value for each call to use different models dynamically, in my get_cmc we actually check if settings.ai_model_id is set, and if not set it to self.ai_model_id, so ideally we first get that (which might have to come in through the decorator constructor, or `self`) and then we check settings to see if it is overridden.

Reply:
This is what is being done here. First, we check if we got the model ID through the settings, if we didn't, we get it through self.ai_model_id (args[0].ai_model_id)

Reply:
> It's tricky because for some, like Azure OpenAI the ai_model_id is the deployment name, so that is a single thing, while for OpenAI you could theoretically override that value for each call to use different models dynamically, in my get_cmc we actually check if settings.ai_model_id is set, and if not set it to self.ai_model_id, so ideally we first get that (which might have to come in through the decorator constructor, or `self`) and then we check settings to see if it is overridden.

I guess this is a more fundamental question: should we give the freedom to set the model for each call that is different than the model already set in the connector?

Reply:
@eavanvalkenburg